code/game/objects/items/weapons/grenades/deconstruction_grenade.dm

```diff
@@ -0,0 +1,288 @@
+	if(choice)
+		to_chat(user, "<span class='notice'>You [choice] to '[outcome]'.</span>")
+
+/obj/item/grenade/deconstruction/proc/check_menu(mob/living/user)
+	if(!istype(user))
+		return FALSE
+	if(user.incapacitated() || !user.Adjacent(src))
+		return FALSE
+	return TRUE
+
+/obj/item/grenade/deconstruction/prime()
+	if(!active) //For if it has been deactivated
+		return
+	unprime()
+	if(get_turf(src) == level_name_to_num(CENTCOMM)) //Safety first.
+		atom_say("Activation denied: Area does not need repossession ")
+		return
+	deconstruct_obj(DECONSTRUCT_ATTEMPTS)
+	addtimer(CALLBACK(src, .proc/reboot), 60 SECONDS)
+	on_cooldown = TRUE
+
+/obj/item/grenade/deconstruction/unprime()
+	..()
+	anchored = FALSE
+	set_light(0)
+
+/obj/item/grenade/deconstruction/proc/deconstruct_obj(loops = 0) //We want structures fully deconstructed, no frames or anything, so multiple go arounds.
+	if(deconstruct_walls)
+		for(var/turf/simulated/wall/W in view(7, src))
+			var/area/A = get_turf(src)
+			if(W.safety_check() && !emagged)
+				continue
+			if(istype(A, /area/engine/supermatter) & !emagged)
+				continue
+			if(loops == DECONSTRUCT_ATTEMPTS) //We only want to do this on first go so we don't display frames disapering
+				new /obj/effect/temp_visual/fadeout(get_turf(W), W)
+			W.dismantle_wall() //Indestructible walls overide this to false
+
+	for(var/obj/structure/S in view(7, src)) //Two runs for objects of structures and machinery, so we don't try to disasemble mechs, or items, or some other strange typepath.
+		var/area/A = get_turf(src)
+		if(S.resistance_flags & INDESTRUCTIBLE)
+			continue
+		if(S.safety_check() && !emagged)
+			continue
+		if(istype(A, /area/engine/supermatter) & !emagged)
+			continue
+		if(istype(S, /obj/structure/window) && !deconstruct_windows)
+			continue
+		if(istype(S, /obj/structure/blob) && !istype(S, /obj/structure/blob/normal)) //I'll let it work on weak blobs, but not on any strong blob structures.
+			continue
+		if(S.invisibility > 30) //for those weird invisible structures that control stuff (looking at you spacevines)
+			continue
+		if(loops == DECONSTRUCT_ATTEMPTS)
+			new /obj/effect/temp_visual/fadeout(get_turf(S), S)
+		S.deconstruct(TRUE)
+
+	for(var/obj/machinery/O in view(7, src))
+		var/area/A = get_turf(src)
+		if(O.resistance_flags & INDESTRUCTIBLE)
+			continue
+		if(O.safety_check() && !emagged)
+			continue
+		if(istype(A, /area/engine/supermatter) & !emagged)
+			continue
+		if(istype(O, /obj/machinery/door) && !deconstruct_doors)
+			continue
+		if(loops == DECONSTRUCT_ATTEMPTS)
+			new /obj/effect/temp_visual/fadeout(get_turf(O), O)
+		O.deconstruct(TRUE)
+
+	if(loops)
+		deconstruct_obj(loops -= 1)
+
+	else if(recycle || consume_all)
+		deconstruct_items()
+
+/obj/item/grenade/deconstruction/proc/deconstruct_items()
+	for(var/obj/item/I in oview(7, src))
+		if(I.resistance_flags & INDESTRUCTIBLE) //No eating objective items, thank you.
+			continue
+		if(length(I.materials) || consume_all)
```

Comment:
Can you make this into an early continue?

Reply:
Yup!